docs/design/tracing.md

```diff
@@ -0,0 +1,148 @@
+
+```kotlin
+interface TraceSpan : Closeable {
+    val id: String
+    val parent: TraceSpan?
+
+    fun child(id: String): TraceSpan
+    fun postEvents(events: Iterable<TraceEvent>)
+}
+```
+
+Spans have an ID (or name) which must be unique among sibling spans within the same parent. Span IDs will generally be
+used by probes to contextualize events.
+
+`TraceSpan` instances are `Closeable` and must be closed when no more events will be emitted to them. Probes may choose
+to batch/aggregate events within a span until a span is closed.
+
+## Trace Event
+
+A `TraceEvent` is the recording of a single event that took place and its associated metadata:
+
+```kotlin
+data class TraceEvent(
+    val level: EventLevel,
+    val sourceComponent: String,
+    val timestamp: Instant,
+    val threadId: String,
+    val data: TraceEventData,
+)
+
+enum class EventLevel { Fatal, Error, Warning, Info, Debug, Trace }
+
+sealed interface TraceEventData {
+  data class Message(val exception: Throwable? = null, val content: () -> Any?) : TraceEventData
+
+  sealed interface Metric : TraceEventData { val metric: String }
+  data class Count<T : Number>(override val metric: String, val count: () -> T) : Metric
+  data class Timespan(override val metric: String, val duration: () -> Duration) : Metric
+}
+```
+
+Trace events occur at different levels (e.g., fatal, info, debug, etc.). These levels may be used by probes to
+include/omit events in their output.
+
+Trace events can be one of three types:
+* `Message`: Typically, a free-form text message used for logging
+* `Count`: The numerical measurement of some value (e.g., results returned, bytes written, etc.)
+* `Timespan`: The temporal measurement of some occurrence (e.g., time elapsed, latency, etc.)
+
+Probes are free to handle these different types of events however they see fit (e.g., they may log some messages,
+aggregate some metrics, ignore some events, etc.).
+
+Event data values (i.e., message text, count values, and timespan durations) are provided as lambdas rather than with
+direct values. This allows probe implementations to skip calculating them in the event they would otherwise be discarded
+(e.g., for events emitted at a level ignored by the probe).
+
+## Trace Probe
+
+A `TraceProbe` is a sink for receiving events from spans. They will typically form a bridge between the SDK's events and
+downstream libraries/frameworks/services which can handle the events. Examples of such downstream systems include Log4j,
+CloudWatch, local files on disk, the console, etc.
+
+SDKs will typically not bundle many implementations of `TraceProbe` themselves. Common probe implementations may be
+available as separate libraries or from third-party sources. Users may implement probes themselves to bridge SDK events
+to whatever downstream systems they desire.
+
+The `TraceProbe` interface is defined as:
+
+```kotlin
+interface TraceProbe {
+    fun postEvents(span: TraceSpan, events: Iterable<TraceEvent>)
```

Comment:
Same concern around singular vs plural. Forcing this into the API feels wrong to me (I'm not going to die on this hill but I feel compelled to bring it up again)

Reply:
As above, will change!